src/_guides/language/effective-dart/design.md

```diff
@@ -1398,10 +1398,64 @@ bool convertToBool(dynamic arg) {
 }
 {% endprettify %}
 
+
+### AVOID using `FutureOr<T>` as a return type.
+
+If a method accepts a `FutureOr<int>`, it is [generous in what it
+accepts][postel]. Users can call the method with either an `int` or a
+`Future<int>`, so they don't need to wrap an `int` in `Future` that you are
+going to unwrap anyway.
+
+[postel]: https://en.wikipedia.org/wiki/Robustness_principle
+
+If you *return* a `FutureOr<int>`, users need to check whether get back an `int`
+or a `Future<int>` before they can do anything useful. (Or they'll just `await`
+the value, effectively always treating it as a `Future`.) Just return a
+`Future<int>`, it's cleaner. It's easier for users to understand that a function
+is either always asynchronous or always synchronous, but a function that can be
+either is hard to use correctly.
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (future-or)"?>
+{% prettify dart %}
+Future<int> triple(FutureOr<int> value) async => (await value) * 3;
+{% endprettify %}
+
+{:.bad-style}
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (future-or)"?>
+{% prettify dart %}
+FutureOr<int> triple(FutureOr<int> value) {
+  if (value is int) return value * 3;
+  return (value as Future<int>).then((v) => v * 3);
+}
+{% endprettify %}
+
+The more precise formulation of this guideline is to *only use `FutureOr<T>` in
+[contravariant][] positions.* Parameters are contravariant and return types are
+covariant. In nested function types, this gets flipped&mdash;if you have a
+parameter whose type is itself a function, then the callback's return type is
+now in contravariant position and the callback's parameters are covariant. This
+means it's OK for a *callback's* type to return `FutureOr<T>`:
+
+[contravariant]: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (future-or-contra)" replace="/FutureOr&lt;S&gt;/[!$&!]/g"?>
```

Comment:
In the `replace` argument, instead of `&lt;S&gt;` use, e.g.:
- `<S\x3E`
- `.S.`
- or the following: 
\<?code-excerpt "misc/..." replace="/FutureOr<b><ins>\S+</ins></b>/[!$&!]/g"?>

Reply:
Done.